handler.go

```diff
@@ -70,7 +70,7 @@ func checkHandleType(t reflect.Type) (*requestType, error) {
 	if t.In(1) != paramsType {
 		return nil, errgo.Newf("second argument is %s, need %s", t.In(1), paramsType)
 	}
-	pt, err := getRequestType(t.In(2))
+	pt, err := getRequestType(preprocessType{reflectType: t.In(2), purpose: purposeUnmarshal})
```

Comment:
I don't think there's any need to specify the purpose when asking for a request type - we may as well do all the work for both marshaling and unmarshaling regardless of when the type was first used. I don't think the overhead will ever be noticeable. So I'd suggest dropping preprocessType and all the purpose stuff.